src/core/symbology/classification/qgsclassificationmethod.cpp

```diff
@@ -0,0 +1,300 @@
+
+  // extra information
+  QDomElement extraElem = doc.createElement( QStringLiteral( "extraInformation" ) );
+  saveExtra( extraElem, context );
+  methodElem.appendChild( extraElem );
+
+  return methodElem;
+}
+
+
+void QgsClassificationMethod::setSymmetricMode( bool enabled, double symmetryPoint, bool astride )
+{
+  mSymmetricEnabled = enabled;
+  mSymmetryPoint = symmetryPoint;
+  mSymmetryAstride = astride;
+}
+
+void QgsClassificationMethod::setLabelPrecision( int precision )
+{
+  // Limit the range of decimal places to a reasonable range
+  precision = qBound( MIN_PRECISION, precision, MAX_PRECISION );
+  mLabelPrecision = precision;
+  mLabelNumberScale = 1.0;
+  mLabelNumberSuffix.clear();
+  while ( precision < 0 )
+  {
+    precision++;
+    mLabelNumberScale /= 10.0;
+    mLabelNumberSuffix.append( '0' );
+  }
+}
+
+QString QgsClassificationMethod::formatNumber( double value ) const
+{
+  if ( mLabelPrecision > 0 )
+  {
+    QString valueStr = QLocale().toString( value, 'f', mLabelPrecision );
+    if ( mLabelTrimTrailingZeroes )
+      valueStr = valueStr.remove( RE_TRAILING_ZEROES );
+    if ( RE_NEGATIVE_ZERO.exactMatch( valueStr ) )
+      valueStr = valueStr.mid( 1 );
+    return valueStr;
+  }
+  else
+  {
+    QString valueStr = QLocale().toString( value * mLabelNumberScale, 'f', 0 );
+    if ( valueStr == QLatin1String( "-0" ) )
+      valueStr = '0';
+    if ( valueStr != QLatin1String( "0" ) )
+      valueStr = valueStr + mLabelNumberSuffix;
+    return valueStr;
+  }
+}
+
+QList<QgsClassificationRange> QgsClassificationMethod::classes( const QgsVectorLayer *vl, const QString &expression, int nclasses )
+{
+  if ( expression.isEmpty() )
+    return QList<QgsClassificationRange>();
+
+  if ( nclasses < 1 )
+    nclasses = 1;
+
+  QList<double> values;
+  double minimum;
+  double maximum;
+
+  int fieldIndex = vl->fields().indexFromName( expression );
+
+  bool ok;
+  if ( mValuesRequired || fieldIndex == -1 )
+  {
+    values = QgsVectorLayerUtils::getDoubleValues( vl, expression, ok );
+    if ( !ok || values.isEmpty() )
+      return QList<QgsClassificationRange>();
+
+    auto result = std::minmax_element( values.begin(), values.end() );
+    minimum = *result.first;
+    maximum = *result.second;
+  }
+  else
+  {
+    minimum = vl->minimumValue( fieldIndex ).toDouble();
+    maximum = vl->maximumValue( fieldIndex ).toDouble();
+  }
+
+  // get the breaks
+  QList<double> breaks = calculateBreaks( minimum, maximum, values, nclasses );
+  breaks.insert( 0, minimum );
+  // create classes
+  return breaksToClasses( breaks );
+}
+
+QList<QgsClassificationRange> QgsClassificationMethod::classes( const QList<double> &values, int nclasses )
+{
+  auto result = std::minmax_element( values.begin(), values.end() );
+  double minimum = *result.first;
+  double maximum = *result.second;
+
+  // get the breaks
+  QList<double> breaks = calculateBreaks( minimum, maximum, values, nclasses );
+  breaks.insert( 0, minimum );
```

Comment:
why is this done here, and not in the subclasses?

Reply:
since the current way it works.
The breaks contains the maximum but not the minimum.
It would make sense to change this, but it might break existing code.